File: templates/python/language-specific-pattern/data-structures.py
"""
Python data-structure patterns for coding interviews.

Use this file when you know the algorithm but need to recall the Pythonic way
to initialize, update, and choose common containers.
"""

from collections import Counter, defaultdict, deque
from heapq import heappop, heappush
from typing import Iterable


# -----------------------------------------------------------------------------
# list: dynamic array, stack, matrix rows
# -----------------------------------------------------------------------------


def list_patterns(nums: list[int]) -> tuple[list[int], list[int], list[list[int]]]:
    # Empty and fixed-size lists.
    values: list[int] = []
    zeros = [0] * len(nums)

    for x in nums:
        values.append(x)  # Amortized O(1)

    # Stack usage: append/pop from the end are O(1).
    stack: list[int] = []
    for x in nums:
        stack.append(x)
    while stack:
        stack.pop()

    # Do not build a matrix with [[0] * cols] * rows: rows share one inner list.
    rows, cols = 2, 3
    matrix = [[0] * cols for _ in range(rows)]

    return values, zeros, matrix


def list_slicing_cost(nums: list[int]) -> list[int]:
    # Slicing copies. Useful, but it is O(k) time and space for slice length k.
    middle = nums[1:-1]
    reversed_copy = nums[::-1]

    # In-place reverse avoids an extra list.
    nums.reverse()

    return middle + reversed_copy


# -----------------------------------------------------------------------------
# dict: hash map, first seen, grouping, counting
# -----------------------------------------------------------------------------


def dict_patterns(words: list[str]) -> tuple[dict[str, int], dict[str, list[str]]]:
    first_index: dict[str, int] = {}
    groups: dict[str, list[str]] = {}

    for i, word in enumerate(words):
        if word not in first_index:
            first_index[word] = i

        key = "".join(sorted(word))
        if key not in groups:
            groups[key] = []
        groups[key].append(word)

    return first_index, groups


def defaultdict_patterns(words: list[str]) -> dict[str, list[str]]:
    # defaultdict calls the factory only when a missing key is read through
    # __getitem__, such as groups[key]. It does not call the factory for get().
    groups: defaultdict[str, list[str]] = defaultdict(list)

    for word in words:
        key = "".join(sorted(word))
        groups[key].append(word)

    # Return as a regular dict when callers do not need missing-key behavior.
    return dict(groups)


def counter_patterns(items: Iterable[str]) -> Counter[str]:
    # Counter is a dict subclass specialized for frequencies. Missing keys read
    # as 0, which is why counts[ch] += 1 works without initialization.
    counts = Counter(items)

    # Most common operations.
    counts.update(["extra"])
    counts.subtract(["extra"])

    # Counter keeps zero/negative counts until deleted or normalized.
    counts += Counter()

    return counts


# -----------------------------------------------------------------------------
# set: membership, deduplication, visited states
# -----------------------------------------------------------------------------


def set_patterns(nums: list[int]) -> tuple[bool, set[int]]:
    seen: set[int] = set()

    for x in nums:
        if x in seen:
            return True, seen
        seen.add(x)

    unique = set(nums)
    return False, unique


# -----------------------------------------------------------------------------
# tuple: immutable compound keys
# -----------------------------------------------------------------------------


def tuple_keys(grid: list[list[int]]) -> set[tuple[int, int]]:
    visited: set[tuple[int, int]] = set()

    for r, row in enumerate(grid):
        for c, value in enumerate(row):
            if value:
                visited.add((r, c))

    return visited


# -----------------------------------------------------------------------------
# deque: real queue for BFS / sliding windows
# -----------------------------------------------------------------------------


def queue_patterns(start: int, graph: dict[int, list[int]]) -> list[int]:
    # deque is a double-ended queue. Both append/pop on either end are O(1),
    # making it the standard Python choice for BFS queues.
    queue = deque([start])
    seen = {start}
    order: list[int] = []

    while queue:
        node = queue.popleft()  # O(1); list.pop(0) would be O(n).
        order.append(node)

        for nei in graph.get(node, []):
            if nei not in seen:
                seen.add(nei)
                queue.append(nei)

    return order


# -----------------------------------------------------------------------------
# heapq: min-heap; negate priorities for max-heap behavior
# -----------------------------------------------------------------------------


def heap_patterns(nums: list[int], k: int) -> list[int]:
    # heapq functions operate on a normal list and maintain the heap invariant:
    # heap[0] is always the smallest item, but the whole list is not sorted.
    heap: list[int] = []

    for x in nums:
        heappush(heap, x)

    smallest = [heappop(heap) for _ in range(min(k, len(heap)))]

    max_heap: list[int] = []
    for x in nums:
        heappush(max_heap, -x)
    largest = [-heappop(max_heap) for _ in range(min(k, len(max_heap)))]

    return smallest + largest


# -----------------------------------------------------------------------------
# Common interview choice guide
# -----------------------------------------------------------------------------


"""
Choose quickly:

- Need order by insertion and random access: list
- Need LIFO stack: list with append/pop
- Need FIFO queue: collections.deque
- Need membership / visited: set
- Need key -> value lookup: dict
- Need missing list/int values: defaultdict(list/int)
- Need frequencies and multiset math: Counter
- Need repeated min/max extraction: heapq
- Need coordinate/state key: tuple

Version notes:

- Python 3.9+: built-in generics like list[int], dict[str, int].
- Python 3.10+: union syntax like int | None.
- Older interviews/environments may expect typing.List and typing.Optional.

Concept explanations:

- `list` is Python's dynamic array. It is good for random access, append/pop at
  the end, stacks, and dense DP tables. Inserting or removing near the front is
  O(n) because elements must shift.
- `[0] * n` is safe for immutable values like integers. `[[0] * cols] * rows`
  is unsafe because each row references the same inner list.
- `dict` is Python's built-in hash map. Keys must be hashable, so integers,
  strings, tuples of hashable values, and frozen dataclasses work; lists and
  dicts do not.
- `dict.setdefault(key, default)` can initialize a missing key inline, but
  `defaultdict` is usually clearer when every missing key uses the same default.
- `defaultdict(list)` means "when a missing key is accessed, create a new list
  for that key." Common factories are `list`, `int`, `set`, and `deque`.
- `Counter(iterable)` builds a frequency map. It is great for anagrams and
  multiset-style comparisons, but a plain `dict` can be clearer when you need
  custom state transitions.
- `set` is a hash table with only keys. Use it for `seen`, `visited`,
  membership checks, and deduplication.
- `tuple` is immutable and hashable if all elements are hashable. This is why
  grid coordinates are usually stored as `(r, c)` rather than `[r, c]`.
- `deque` is from `collections`, not the language core. It is the right queue
  abstraction for BFS because `popleft()` is O(1).
- `heapq` is a module, not a heap class. You pass a list to `heappush` and
  `heappop`. Python only provides a min-heap directly; use negative priorities
  or reversed comparison fields for max-heap behavior.
"""

File: templates/golang/language-specific-pattern/data-structures.go
// Go data-structure patterns for coding interviews.
//
// Go's core interview containers are arrays, slices, and maps. Stacks and
// queues are normally slices; sets are normally maps with empty struct values.
package languagepatterns

// SlicePatternsData demonstrates creation, copying, appending, and deletion.
func SlicePatternsData(nums []int) (copied, inserted, deleted []int) {
	// make([]int, length, capacity): length elements are immediately indexable.
	buffer := make([]int, 0, len(nums)+1)
	buffer = append(buffer, nums...)

	// copy duplicates elements. Plain assignment would copy only the slice
	// header, leaving both slices backed by the same array.
	copied = make([]int, len(nums))
	copy(copied, nums)

	// Insert before index 1 without relying on a particular spare capacity.
	inserted = append([]int(nil), nums...)
	index := min(1, len(inserted))
	inserted = append(inserted, 0)
	copy(inserted[index+1:], inserted[index:])
	inserted[index] = 99

	// Delete index 0 while preserving order. Clear the now-unused tail slot
	// when element references should become collectible promptly.
	deleted = append([]int(nil), buffer...)
	if len(deleted) > 0 {
		copy(deleted, deleted[1:])
		deleted[len(deleted)-1] = 0
		deleted = deleted[:len(deleted)-1]
	}

	return copied, inserted, deleted
}

// MatrixData allocates each row separately. This is the conventional shape for
// a rectangular interview grid.
func MatrixData(rows, cols int) [][]int {
	grid := make([][]int, rows)
	for row := range grid {
		grid[row] = make([]int, cols)
	}
	return grid
}

// SliceViewData demonstrates that slicing does not copy. Changes through view
// are visible in nums until an append moves view to a different backing array.
func SliceViewData(nums []int, start, end int) []int {
	view := nums[start:end]
	if len(view) > 0 {
		view[0] = -1
	}
	return view
}

// StackData uses the right end of a slice, where push/pop are amortized O(1).
func StackData(values []int) []int {
	stack := make([]int, 0, len(values))
	for _, value := range values {
		stack = append(stack, value)
	}

	result := make([]int, 0, len(stack))
	for len(stack) > 0 {
		last := len(stack) - 1
		result = append(result, stack[last])
		stack = stack[:last]
	}
	return result
}

// QueueData uses a head index so removing the front does not shift O(n)
// elements. For a long-lived queue, compact occasionally so old entries and a
// large backing array are not retained forever.
func QueueData(values []int) []int {
	queue := append([]int(nil), values...)
	order := make([]int, 0, len(queue))
	for head := 0; head < len(queue); head++ {
		order = append(order, queue[head])
	}
	return order
}

// BFSData shows the standard slice queue plus a map-backed visited set.
func BFSData(start int, graph map[int][]int) []int {
	queue := []int{start}
	seen := map[int]struct{}{start: {}}
	order := make([]int, 0, len(graph))

	for head := 0; head < len(queue); head++ {
		node := queue[head]
		order = append(order, node)
		for _, neighbor := range graph[node] {
			if _, exists := seen[neighbor]; exists {
				continue
			}
			seen[neighbor] = struct{}{}
			queue = append(queue, neighbor)
		}
	}
	return order
}

// FrequencyData demonstrates map zero values and the comma-ok lookup.
func FrequencyData(words []string, query string) (map[string]int, int, bool) {
	counts := make(map[string]int, len(words))
	for _, word := range words {
		counts[word]++ // A missing int value reads as 0.
	}

	count, present := counts[query]
	return counts, count, present
}

// GroupData relies on append working with a nil slice returned by a missing map
// key. No separate "initialize group" branch is needed.
func GroupData(words []string) map[byte][]string {
	groups := make(map[byte][]string)
	for _, word := range words {
		if word == "" {
			continue
		}
		key := word[0]
		groups[key] = append(groups[key], word)
	}
	return groups
}

// SetData uses map[int]struct{} because struct{} occupies zero bytes as a map
// value. map[int]bool can be convenient when false has useful meaning.
func SetData(nums []int) (map[int]struct{}, bool) {
	seen := make(map[int]struct{}, len(nums))
	hasDuplicate := false
	for _, value := range nums {
		if _, exists := seen[value]; exists {
			hasDuplicate = true
		}
		seen[value] = struct{}{}
	}
	return seen, hasDuplicate
}

// ArrayValueData shows that an array's length is part of its type and assigning
// an array copies every element.
func ArrayValueData(input [3]int) ([3]int, [3]int) {
	copied := input
	copied[0] = -1
	return input, copied
}

/*
Container choice guide:

- Random access, stack, dense DP table: []T.
- FIFO queue: []T with a head index for interview-sized workloads.
- Key to value, frequency, grouping: map[K]V.
- Membership/visited: map[T]struct{}.
- Fixed small value whose length is meaningful: [N]T.

Important semantics:

- A slice is a descriptor (pointer, length, capacity), not an owning list.
  Assignment shares its backing array; copy or append to a fresh slice to clone.
- append returns the resulting slice and may allocate. Always retain its return
  value. A callee that appends cannot update the caller's slice header unless it
  returns the slice or receives *[]T.
- nil slices are safe to len, range, and append. Reading a nil map is safe, but
  assigning a key in a nil map panics.
- Map keys must be comparable. Slices, maps, and functions cannot be keys.
- Map iteration order is unspecified and can differ between iterations.
- delete on a missing key is safe. clear(map) removes all entries in Go 1.21.
*/
